storage/src/storage_backend.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) SimpleStaking and Tezedge Contributors
+// SPDX-License-Identifier: MIT
+
+use crate::persistent::database::{DBError, RocksDBStats};
+use failure::Fail;
+use serde::Serialize;
+use std::collections::HashSet;
+use std::mem;
+
+use crate::merkle_storage::{ContextValue, EntryHash};
+
+pub fn size_of_vec<T>(v: &Vec<T>) -> usize {
+    mem::size_of::<Vec<T>>() + mem::size_of::<T>() * v.capacity()
+}
+
+#[derive(Debug, Fail)]
+pub enum StorageBackendError {
+    #[fail(display = "RocksDB error: {}", error)]
+    RocksDBError { error: rocksdb::Error },
+    #[fail(display = "Column family {} is missing", name)]
+    MissingColumnFamily { name: &'static str },
+    #[fail(display = "Backend Error")]
+    BackendError,
+    #[fail(display = "SledDB error: {}", error)]
+    SledDBError { error: sled::Error },
+    #[fail(display = "Guard Poison {} ", error)]
+    GuardPoison { error: String },
+    #[fail(display = "Serialization error: {:?}", error)]
+    SerializationError { error: bincode::Error },
+    #[fail(display = "DBError error: {:?}", error)]
+    DBError { error: DBError },
+}
+
+impl From<rocksdb::Error> for StorageBackendError {
+    fn from(error: rocksdb::Error) -> Self {
+        StorageBackendError::RocksDBError { error }
+    }
+}
+
+impl From<sled::Error> for StorageBackendError {
+    fn from(error: sled::Error) -> Self {
+        StorageBackendError::SledDBError { error }
+    }
+}
+
+impl From<DBError> for StorageBackendError {
+    fn from(error: DBError) -> Self {
+        StorageBackendError::DBError { error }
+    }
+}
+
+impl From<bincode::Error> for StorageBackendError {
+    fn from(error: bincode::Error) -> Self {
+        StorageBackendError::SerializationError { error }
+    }
+}
+
+impl slog::Value for StorageBackendError {
+    fn serialize(
+        &self,
+        _record: &slog::Record,
+        key: slog::Key,
+        serializer: &mut dyn slog::Serializer,
+    ) -> slog::Result {
+        serializer.emit_arguments(key, &format_args!("{}", self))
+    }
+}
+
+pub trait StorageBackend: Send + Sync {
+    fn is_persisted(&self) -> bool;
+    fn get(&self, key: &EntryHash) -> Result<Option<ContextValue>, StorageBackendError>;
+    fn put(&mut self, key: EntryHash, value: ContextValue) -> Result<bool, StorageBackendError>;
```

Comment:
`put` and `merge` take ownership of the key while the others borrow from it, why is that?

Reply:
fixed